src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java

```diff
@@ -0,0 +1,349 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java.migrate.lombok;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.EqualsAndHashCode;
+import lombok.Value;
+import org.openrewrite.Cursor;
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Option;
+import org.openrewrite.Preconditions;
+import org.openrewrite.Recipe;
+import org.openrewrite.TreeVisitor;
+import org.openrewrite.internal.lang.Nullable;
+import org.openrewrite.java.AnnotationMatcher;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.JavaTemplate;
+import org.openrewrite.java.RemoveAnnotationVisitor;
+import org.openrewrite.java.search.UsesJavaVersion;
+import org.openrewrite.java.tree.Expression;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.JavaType;
+import org.openrewrite.java.tree.Statement;
+import org.openrewrite.java.tree.TypeUtils;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static java.util.Objects.requireNonNull;
+
+@Value
+@EqualsAndHashCode(callSuper = false)
+public class LombokValueToRecord extends Recipe {
+
+    @Option(displayName = "useExactToString",
+            description = "When set the toString format from Lombok is used in the migrated record.")
+    boolean useExactToString;
+
+    @JsonCreator
+    public LombokValueToRecord(final @JsonProperty("useExactToString") boolean useExactToString) {
+        this.useExactToString = useExactToString;
+    }
+
+    @Override
+    public String getDisplayName() {
+        return "Convert Value class to Record";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Convert Lombok Value annotated classes to standard Java Record classes in Java 17 or higher.";
+    }
+
+    @Override
+    public Set<String> getTags() {
+        return Collections.singleton("lombok");
+    }
+
+    @Override
+    public Duration getEstimatedEffortPerOccurrence() {
+        return Duration.ofMinutes(5);
+    }
+
+    @Override
+    public boolean causesAnotherCycle() {
+        return true;
+    }
+
+    @Override
+    public int maxCycles() {
+        return 2;
+    }
+
```

Comment:
I'm doubtful if we need these overrides; were these added intentionally & necessary for these replacements?
```suggestion
```

Reply:
Yes currently these overrides are required since the recipe runs in two phases

- First run: adjust the value classes to records and keep track of the members 
- Second run: adjust the references to getters of the adjusted classes

I haven't found another way to solve the issue of adjusting the references in the same cycle because from what I have read in the documentation the recipe is run in parallel when executed outside of a test environment.

If there is a solution how to solve this with only one run I would be really interested in it.

Reply:
Enter `ScanningRecipe`: https://docs.openrewrite.org/concepts-explanations/recipes#scanning-recipes

Recipes extending this class will run in two phases: The first phase collects data across all sources and the second phase performs the required modifcations.

Reply:
Thanks for the hint, I refactored the implementation so that a ScanningRecipe is used.